### PR TITLE
Added `literal` param to string-replace functions, optimized `replace` performance in small-string regime (30-80% faster)

### PR DESCRIPTION
**Pull request by alexander-beedie**
This is a comprehensive revisit of a previously proposed patch; buckle-up for the fun details - "there will be graphs"  ;)

TLDR
-----
* Python: string `replace` and `replace_all` methods have a new `literal` param.
* Rust: added `replace_literal` and `replace_literal_all` methods.
* String `replace` ops for strings <= 32 chars are 30-80% faster for literal values.
* String `contains` op speed restored for larger strings (> 96 chars) when searching for literals (regex not impacted).
* Nothing (in any string-size regime) got slower.

Deeper-dive
-----
Originally it was thought that the regex engine would display overhead resulting in slower string find/replace for literals vs a simple `str.replace` op, and early testing appeared to confirm this. However... I had a few reservations, so I put together some benchmarking functionality to investigate performance across multiple dimensions, specifically:

* Varied underlying string length from small (8) to large (65,536).
* Position within the string of the target literal (front, back, random).
* Percentage of strings in the dataframe/series that would match & get replaced.

Test DataFrame was composed of 5 cols, each containing 100,000 strings.
This analysis showed some very interesting results...
```
x-axis: % of strings replaced in the frame (from 5% to 100%).
y-axis: time taken to complete replace op (in secs).
```
* The `str.replace` op _is_ faster than regex across the entire replace % domain... but _only_ for <= 32 chars.
* There is an immediate fall-off from 32-33 chars; str.replace is faster only when replacing > 35% of the strings.
  
  image

* The str.replace gradient across different string-size regimes varies; regex gradient is relatively stable:

  image

  * Inference - regex is much faster at identifying literal null-matches over larger strings.
    Corroboration: aho-corasick / memchr usage within regex lib source for literal matches -
    https://github.com/rust-lang/regex/blob/master/src/literal/imp.rs
  *  Inference - str.replace is faster to replace once a match is found.
     Corroboration - the narrowing gap between regex/str.replace as a higher percentage of matches are found.

What this PR does
-----
Based on the amassed data, `str.replace` is used for literal replace only in the region where it is guaranteed to be faster in all cases (<= 32 chars target string). In this region it is, however, _much_ faster - between 30-80%, depending on how many strings will be replaced (the more, the better).

In all other cases the existing regex replace is used. So, the patch will _never_ be slower than the current code, but in the common short-string regime it can be a lot faster.

The previous `contains` patch for literals contains a similar, but not identical, fall-off; I have modified that function similarly so that `str.contains` is invoked for literal search when the target string is <= 96 chars - otherwise the regex lib is used.

Future work
-----
* Potential hybrid strategy, using regex find_iter in conjunction with str.replace to get "best of both worlds" performance.
* I experimented with `cow_utils.cow_replacen` instead of `str.replacen`, but saw no gains in the small-string regime. Will return to this sometime, as it seems gains _should_ be possible.
* Modifying existing str buffer for literal replaces of the same size (buffer re-use).

Additional note
-----
* regex lib is faster across all regimes for `replace_all` functionality, so only added literal param option.
* Updating string functionality requires testing across multiple regimes to validate performance characteristics!

### CODE REVIEW

**Comment by ritchie46** [comment]
Very nice work! Could you keep your benchmark setup? I want to follow up with a better implementation of `str::replace`. One that returns `Cow<str>` instead of `String`.

Currently that method always allocates a new `String` where it very common that most strings don't match the pattern.

**Comment by alexander-beedie** [comment]
> Very nice work! Could you keep your benchmark setup? 

> Very nice work! Could you keep your benchmark setup?

No problem; will tidy it up a bit and send it over (or make it available as a gist?) 👍 

> I want to follow up with a better implementation of `str::replace`. One that returns `Cow<str>` instead of `String`.

Yup. I was a bit surprised that `cow_utils` didn't make an obvious difference; may have overlooked something, hmm...
